src/clp_ffi_js/ir/IrStreamReader.hpp

```diff
@@ -0,0 +1,113 @@
+#ifndef CLP_FFI_JS_IR_IR_STREAM_READER_HPP
+#define CLP_FFI_JS_IR_IR_STREAM_READER_HPP
+
+#include <Array.hpp>
+#include <cstddef>
+#include <memory>
+#include <optional>
+#include <streaming_compression/zstd/Decompressor.hpp>
+#include <vector>
+
+#include <clp/ir/types.hpp>
+#include <clp/TimestampPattern.hpp>
+#include <emscripten/val.h>
+
+#include <clp_ffi_js/ir/LogEventWithLevel.hpp>
+#include <clp_ffi_js/ir/StreamReader.hpp>
+#include <clp_ffi_js/ir/StreamReaderDataContext.hpp>
+
+namespace clp_ffi_js::ir {
+using clp::ir::four_byte_encoded_variable_t;
+
+/**
+ * Mapping between an index in the filtered log events collection to an index in the unfiltered
+ * log events collection.
+ */
+using FilteredLogEventsMap = std::optional<std::vector<size_t>>;
+
+/**
+ * Class to deserialize and decode Zstandard-compressed CLP IRv1 streams as well as format decoded
+ * log events.
+ */
+class IrStreamReader : public StreamReader {
```

Comment:
How about calling this `TextIrStreamReader` and `StreamReader` -> `IrStreamReader`?

Reply:
How about inserting `Ir` before `Stream`---`UnstructuredIrStreamReader` and `StructuredIrStreamReader`? (Once the new IR format supports backward compatibility, hopefully we can revert back to a single `IrStreamReader` class.)

> since he wants to one day have an ffi for single file archive.

I see. I guess we can keep `StreamReader` then.

> Text seems a bit vague

It's how we differentiate unstructured and structured for the casual user. "Text" for unstructured and "JSON" for structured.